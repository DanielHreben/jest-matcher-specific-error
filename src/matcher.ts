type MatcherContext = jest.MatcherContext;
type CustomMatcherResult = jest.CustomMatcherResult;

function errorToObject(error: Error) {
  const { stack, ...publicFields } = error;

  return {
    class: error.constructor,
    message: error.message,
    publicFields,
  };
}

export function toMatchError(
  this: MatcherContext,
  gotError: Error,
  expectedError: Error
): CustomMatcherResult {
  const got = errorToObject(gotError);
  const expected = errorToObject(expectedError);

  const diff: string[] = [];

  if (got.class !== expected.class) {
    diff.push(
      "Error class is not the same:",
      this.utils.diff(got.class, expected.class) as string
    );
  }

  if (got.message !== expected.message) {
    diff.push(
      "Error message is not the same:",
      this.utils.diff(got.message, expected.message) as string
    );
  }

  if (!this.equals(got.publicFields, expected.publicFields)) {
    diff.push(
      "Error public fields is not the same:",
      this.utils.diff(got.publicFields, expected.publicFields) as string
    );
  }

  return {
    pass: diff.length === 0,
    message: () => diff.join("\n"),
  };
}
