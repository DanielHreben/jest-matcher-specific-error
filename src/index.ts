import { toMatchError } from "./matcher";

const jestExpect = ((global as NodeJS.Global) && { expect }).expect;

if (jestExpect !== undefined) {
  jestExpect.extend({ toMatchError });
} else {
  throw new Error(
    [
      "Unable to find Jest's global expect.",
      "Please check you have added jest-extended correctly to your jest configuration.",
      "See https://github.com/jest-community/jest-extended#setup for help.",
    ].join("\n")
  );
}
