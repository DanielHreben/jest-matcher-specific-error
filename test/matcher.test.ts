import "../src";

class SpecificError extends Error {
  constructor(message: string, readonly other: any = {}) {
    super(message);
  }
}

class SpecificError2 extends Error {
  constructor(message: string, readonly other: any = {}) {
    super(message);
  }
}

function throwError(err: Error) {
  return () => {
    throw err;
  };
}

describe("matcher", () => {
  describe("error factory", () => {
    it("should be successful if the error factory result matches", () => {
      expect(throwError(new SpecificError("test"))).toMatchError(
        new SpecificError("test")
      );
    });

    it("should throw error if different types", () => {
      expect(() => {
        expect(throwError(new SpecificError2("test"))).toMatchError(
          new SpecificError("test")
        );
      }).toThrowErrorMatchingInlineSnapshot(`
        "Error class is not the same:
        - Expected
        + Received

        - [Function SpecificError]
        + [Function SpecificError2]"
      `);
    });

    it("should throw error if different messages", () => {
      expect(() => {
        expect(throwError(new SpecificError("test"))).toMatchError(
          new SpecificError("test2")
        );
      }).toThrowErrorMatchingInlineSnapshot(`
        "Error message is not the same:
        - Expected
        + Received

        - test2
        + test"
      `);
    });

    it("should throw error if different properties", () => {
      expect(() => {
        expect(throwError(new SpecificError("test"))).toMatchError(
          new SpecificError("test", false)
        );
      }).toThrowErrorMatchingInlineSnapshot(`
        "Error public fields is not the same:
        - Expected
        + Received

          Object {
        -   "other": false,
        +   "other": Object {},
          }"
      `);
    });

    it("should fail if no error is thrown", () => {
      expect(() => {
        expect(() => true).toMatchError(new SpecificError("test"));
      }).toThrowErrorMatchingInlineSnapshot(
        `"Expected to receive an error, but no error was thrown"`
      );
    });
  });

  describe("rejected promises", () => {
    it("should be successful if the error factory result matches", async () => {
      await expect(
        Promise.reject(new SpecificError("test"))
      ).rejects.toMatchError(new SpecificError("test"));
    });

    it("should throw an error if error factory result does not match", async () => {
      await expect(async () => {
        await expect(
          Promise.reject(new SpecificError("test"))
        ).rejects.toMatchError(new SpecificError2("test"));
      }).rejects.toThrowErrorMatchingInlineSnapshot(`
              "Error class is not the same:
              - Expected
              + Received

              - [Function SpecificError2]
              + [Function SpecificError]"
            `);
    });
  });
});
