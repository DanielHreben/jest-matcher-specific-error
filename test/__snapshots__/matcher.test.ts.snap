// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matcher error factory should fail if no error is thrown 1`] = `"Expected to receive an error, but no error was thrown"`;

exports[`matcher error factory should throw error if different messages 1`] = `
"Error message is not the same:
- Expected
+ Received

- test
+ test2"
`;

exports[`matcher error factory should throw error if different properties 1`] = `
"Error public fields is not the same:
- Expected
+ Received

  Object {
-   \\"other\\": Object {},
+   \\"other\\": false,
  }"
`;

exports[`matcher error factory should throw error if different types 1`] = `
"Error class is not the same:
- Expected
+ Received

- [Function SpecificError2]
+ [Function SpecificError]"
`;

exports[`matcher rejected promises should throw an error if error factory result does not match 1`] = `
"Error class is not the same:
- Expected
+ Received

- [Function SpecificError]
+ [Function SpecificError2]"
`;
